core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/GroupNodeConstructor.java

```diff
@@ -36,6 +180,128 @@ AbstractNode apply(int groupStoreIndex, TupleLifecycle<Tuple_> nextNodesTupleLif
 
     }
 
+    @FunctionalInterface
+    interface GroupBy0Mapping1CollectorNodeBuilder<CollectorA_, Tuple_ extends AbstractTuple> {
+        AbstractNode build(int groupStoreIndex, int undoStoreIndex,
+                CollectorA_ collector,
+                TupleLifecycle<Tuple_> nextNodesTupleLifecycle, int outputStoreSize, EnvironmentMode environmentMode);
+    }
+
+    @FunctionalInterface
+    interface GroupBy0Mapping2CollectorNodeBuilder<CollectorA_, CollectorB_, Tuple_ extends AbstractTuple> {
+        AbstractNode build(int groupStoreIndex, int undoStoreIndex,
+                CollectorA_ collectorA,
+                CollectorB_ collectorB,
+                TupleLifecycle<Tuple_> nextNodesTupleLifecycle, int outputStoreSize, EnvironmentMode environmentMode);
+    }
+
+    @FunctionalInterface
+    interface GroupBy0Mapping3CollectorNodeBuilder<CollectorA_, CollectorB_, CollectorC_, Tuple_ extends AbstractTuple> {
+        AbstractNode build(int groupStoreIndex, int undoStoreIndex,
+                CollectorA_ collectorA,
+                CollectorB_ collectorB,
+                CollectorC_ collectorC,
+                TupleLifecycle<Tuple_> nextNodesTupleLifecycle, int outputStoreSize, EnvironmentMode environmentMode);
+    }
+
+    @FunctionalInterface
+    interface GroupBy0Mapping4CollectorNodeBuilder<CollectorA_, CollectorB_, CollectorC_, CollectorD_, Tuple_ extends AbstractTuple> {
```

Comment:
One thing just crossed my mind... if we don't need `Collector_` to extend `...ConstraintCollector`, do we need `Tuple_` to extend `AbstractTuple`?
If not, even more boilerplate to remove right here.

Reply:
It should not be needed, but it allows for type safety. Unlike Tuple(s), the ConstraintCollector interfaces do not have a common superinterface that we can use here.

Reply:
Fair enough.